frontend/src/Components/CardSelector.tsx

```diff
@@ -4,25 +4,31 @@ import classes from './CardSelector.module.css';
 import { IconCoffee } from './IconCoffee';
 import { connectToWebSocket } from './WebSocket';
 import { IconObserver } from './IconObserver';
-import { BUTTON_OBSERVER, VOTE_COFFEE, VOTE_OBSERVER } from '../constants';
+import { BUTTON_OBSERVER, VOTE_COFFEE, VOTE_NOTE_VOTED, VOTE_OBSERVER } from '../constants';
 
 const ProtoCardSelector = ({ socket }: { socket: WebSocketApi }) => {
   const selectedCard = socket.state.votes[socket.loginData.user];
+
   return (
     <>
       <div class={classes.cardCollection}>
-        {socket.state.scale.map((cardValue) => (
-          <button
-            key={cardValue}
-            class={classNames([
-              classes.largeCard,
-              { [classes.selected]: selectedCard === cardValue },
-            ])}
-            onClick={() => socket.setVote(cardValue)}
-          >
-            {cardValue === VOTE_COFFEE ? <IconCoffee /> : cardValue}
-          </button>
-        ))}
+        {socket.state.scale.map((cardValue) => {
+          return (
+            <button
+              key={cardValue}
+              class={selectedCard === cardValue ? classNames([classes.largeCard, classes.selected]) : classNames([classes.largeCard])}
+              onClick={() => {
+                if (selectedCard === cardValue) {
```

Comment:
We can refactor this as
```
selectedCard === cardValue ? socket.setVote(VOTE_NOT_VOTED) : socket.setVote(cardValue)
```
or even
```
socket.setVote(selectedCard === cardValue ? VOTE_NOT_VOTED : cardValue)
```
if you prefer

Reply:
I would extract the onClick handler above the return statement

Reply:
Personally, I prefer the second, concise version.

If you extract things, I would extract the boolean `selectedCard === cardValue` to e.g. `isSelected`. Then maybe it can be argued that the rest could also remain inline as it will be more compact:

```js
          const isSelected = selectedCard === cardValue;
          return (
            <button
              key={cardValue}
              class={classNames([classes.largeCard, { [classes.selected]: isSelected }])}
              onClick={() => socket.setVote(isSelected ? VOTE_NOTE_VOTED : cardValue)}
            >
              {cardValue === VOTE_COFFEE ? <IconCoffee /> : cardValue}
            </button>
```

But more extraction is definitely ok as well.

Reply:
Small inconsistency: The observer button cannot be deselected.

Reply:
Thinking about this more, maybe rather extract a function that generates buttons outside the entire component. The only difference between the observer and the other cards is the "largeCard" class. So something like

```js
import { JSX } from 'preact';

const SPECIAL_ICONS: { [value in CardValue]?: JSX.Element } = {
  [VOTE_OBSERVER]: <IconObserver />,
  [VOTE_COFFEE]: <IconCoffee />,
};

const getCard = (
  cardValue: CardValue,
  isSelected: boolean,
  isObserver: boolean,
  socket: WebSocketApi
) => (
  <button
    key={cardValue}
    class={classNames({
      [classes.buttonObserver]: isObserver,
      [classes.largeCard]: !isObserver,
      [classes.selected]: isSelected,
    })}
    onClick={() => socket.setVote(isSelected ? VOTE_NOTE_VOTED : cardValue)}
  >
    {SPECIAL_ICONS[cardValue] || cardValue}
    {isObserver && <div class={classes.buttonObserverText}>{BUTTON_OBSERVER}</div>}
  </button>
);
```

Then the actual component will become rather small.

Reply:
Alternatively, one could extract some logic back to the usage site

```js
const getCard = (
  cardValue: CardValue,
  isSelected: boolean,
  additionalClass: string,
  additionalElement: JSX.Element | null,
  socket: WebSocketApi
) => (
  <button
    key={cardValue}
    class={classNames([{ [classes.selected]: isSelected }, additionalClass])}
    onClick={() => socket.setVote(isSelected ? VOTE_NOTE_VOTED : cardValue)}
  >
    {SPECIAL_ICONS[cardValue] || cardValue}
    {additionalElement}
  </button>
);

const ProtoCardSelector = ({ socket }: { socket: WebSocketApi }) => {
  const selectedCard = socket.state.votes[socket.loginData.user];

  return (
    <>
      <div class={classes.cardCollection}>
        {socket.state.scale.map((cardValue) =>
          getCard(cardValue, selectedCard === cardValue, classes.largeCard, null, socket)
        )}
      </div>
      {getCard(
        VOTE_OBSERVER,
        selectedCard === VOTE_OBSERVER,
        classes.buttonObserver,
        <div class={classes.buttonObserverText}>{BUTTON_OBSERVER}</div>,
        socket
      )}
    </>
  );
};
```

Opinions?

Reply:
Last, thinking a little more about `classNames`, maybe it is easier instead of the mixed object form to make use of the fact that `classNames` prunes undefined values and write it like this:

```
classNames([isSelected && classes.selected, additionalClass]
```

Reply:
I personally prefer the version of `getCard` that takes only four arguments. It makes the code easier to read, since that way `getCard` abstracts more from the source code. I think that I'd rather read this:
```
{getCard(VOTE_OBSERVER, selectedCard === VOTE_OBSERVER, true, socket)}
```
instead of this:
```
{getCard(VOTE_OBSERVER, selectedCard === VOTE_OBSERVER,  classes.buttonObserver,
        <div class={classes.buttonObserverText}>{BUTTON_OBSERVER}</div>,
        socket
)}
```

Reply:
@andreashille regarding your comment:
> I would extract the onClick handler above the return statement
I did agree with it in the original implementation, but I think that it is no longer needed after the refactoring suggested by @lukastaegert . However, please have a look at the new implementation and let me know if you still believe that extracting the handler would be best